File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/support/Explanation.java
package fr.myprysm.vertx.elasticsearch.action.support;

import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@DataObject(generateConverter = true)
public class Explanation {

    private boolean match;
    private float value;
    private String description;
    private List<Explanation> details;

    public Explanation(Explanation other) {
        match = other.match;
        value = other.value;
        description = other.description;
        details = other.details;
    }

    /**
     * Build a new <code>Explanation</code> from a <code>JsonObject</code>.
     *
     * @param json the <code>JsonObject</code>
     */
    public Explanation(JsonObject json) {
        ExplanationConverter.fromJson(json, this);
    }


    /**
     * Transforms the <code>Explanation</code> into a <code>JsonObject</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = new JsonObject();
        ExplanationConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/support/Explanation.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.support

import fr.myprysm.vertx.elasticsearch.action.support.Explanation

fun Explanation(
  description: String? = null,
  details: Iterable<fr.myprysm.vertx.elasticsearch.action.support.Explanation>? = null,
  match: Boolean? = null,
  value: Float? = null): Explanation = fr.myprysm.vertx.elasticsearch.action.support.Explanation().apply {

  if (description != null) {
    this.setDescription(description)
  }
  if (details != null) {
    this.setDetails(details.toList())
  }
  if (match != null) {
    this.setMatch(match)
  }
  if (value != null) {
    this.setValue(value)
  }
}

